### PR TITLE
Clippy Book Chapter Updates Reborn: Writing tests

### PR DESCRIPTION
**Pull request by blyxyas**
This PR adds  a new chapter to the book: "Writing tests". The changes have been mainly done from reviews from #9426 and some minor re-writes.

## Notes

- We still need to check that the `git status`es are correct, as `cargo dev new_lint` changed a lot since 2022.
- Requires #10598: Link to "Emitting Lints" where I flagged with `FIXME:`.
- To talk about the whole project, please use the tracking issue for the project #10597 (It also contains a timeline, discussions and more information)

changelog: Add a new "Writing tests" chapter to the book
r? @flip1995

### CODE REVIEW

File: book/src/development/writing_tests.md [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.48s
+```
+
+This is normal. After all, we wrote a bunch of Rust code but we haven't really
+implemented any logic for Clippy to detect `foo` functions and emit a lint.
+
+As we gradually implement our lint logic, we will keep running this UI test command.
+Clippy will begin outputting information that allows us to check if the output is
+turning into what we want it to be.
+
```
Comment (dswij) [inline book/src/development/writing_tests.md:91]: I think we can add some simple lint logic and show the output here

File: book/src/development/writing_tests.md [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+	tests/ui-cargo/foo_categories/
+```
+
+## Rustfix Tests
+
+If the lint you are working on is making use of structured suggestions, the test
+file should include a `// run-rustfix` comment at the top.
+
+Structured suggestions tell a user how to fix or re-write certain code that has been linted, they are usually linted
+with [`span_lint_and_sugg`](https://doc.rust-lang.org/beta/nightly-rustc/clippy_utils/diagnostics/fn.span_lint_and_sugg.html).
```
Comment (dswij) [inline book/src/development/writing_tests.md:168]: Maybe we can mention a bit about `Applicability` in relation to rustfix.

Comment (blyxyas) [reply book/src/development/writing_tests.md:168]: I think leaving a link and "We'll talk about this more in depth in a later chapter" would be the right solution.

File: book/src/development/writing_tests.md (hunk 1 above)
Comment (blyxyas) [reply book/src/development/writing_tests.md:88]: Added.

File: book/src/development/writing_tests.md [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+Developing lints for Clippy is a Test-Driven Development (TDD) process because
+our first task before implementing any logic for a new lint is to write some test cases.
+
+## Develop Lint with Tests
+
+It is not necessarily that every Clippy developer enjoys having the compiler screaming
+test failures in their face. Some embrace TDD *while* others learn to peacefully co-exist with it.
+
+When we develop Clippy, we enter a complex and chaotic realm full of 
+programmatic issues, stylistic errors, illogical code and non-adherence to convention.
```
[the remark this answers is not in the document]
Comment (blyxyas) [reply book/src/development/writing_tests.md]: Fixed

File: book/src/development/writing_tests.md (hunk 2 above)
Comment (blyxyas) [reply book/src/development/writing_tests.md:199]: Fixed.

**Review by flip1995 — APPROVED** [review]
This LGTM overall. So I just did some cleanup (formatting) and added some documentation about `ui-toml` tests.

File: book/src/development/writing_tests.md [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+Update the file with some examples to get started:
+
+```rust
+#![warn(clippy::foo_functions)]
+// Impl methods
+struct A;
+impl A {
+    pub fn fo(&self) {}
+    pub fn foo(&self) {}
+    pub fn food(&self) {}
```
Comment (flip1995) [inline book/src/development/writing_tests.md:45]: Can you update this with the soon to be added error annotations, please? #11249

File: book/src/development/writing_tests.md [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+   |    ^^^
+
+error: aborting due to 3 previous errors
+```
+
+Note the *failures* label at the top of the fragment, we'll get rid of it (saving this output) in the next section.
+
+> _Note:_ You can run multiple test files by specifying a comma separated list:
+> `TESTNAME=foo_functions,bar_methods,baz_structs`.
+### `cargo dev bless`
```
Comment (flip1995) [inline book/src/development/writing_tests.md:129]:
```suggestion

### `cargo bless`
```

File: book/src/development/writing_tests.md [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+
+Once we are satisfied with the output, we need to run this command to
+generate or update the `.stderr` file for our lint:
+
+```sh
+$ TESTNAME=foo_functions cargo uitest
+# (Output is as we want it to be)
+$ cargo dev bless
```
Comment (flip1995) [inline book/src/development/writing_tests.md:137]: I think 

```suggestion
$ TESTNAME=foo_functions cargo uibless
```

is enough now.

File: book/src/development/writing_tests.md [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+
+## Clippy UI Tests
+
+We use **UI tests** for testing in Clippy.
+These UI tests check that the output of Clippy is exactly as we expect it to be.
+Each test is just a plain Rust file that contains the code we want to check.
+
+The output of Clippy is compared against a `.stderr` file.
+Note that you don't have to create this file yourself.
+We'll get to generating the `.stderr` files with the command [`cargo dev bless`](#cargo-dev-bless) (seen later on).
```
Comment (flip1995) [inline book/src/development/writing_tests.md]:
```suggestion
We'll get to generating the `.stderr` files with the command [`cargo bless`](#cargo-bless) (seen later on).
```

File: book/src/development/writing_tests.md [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+# (Output is as we want it to be)
+$ cargo dev bless
+```
+
+This write the emitted lint suggestions and fixes to the `.stderr` file,
+with the reason for the lint, suggested fixes, and line numbers, etc.
+
+> _Note:_ we need to run `TESTNAME=foo_functions cargo uitest` every time before we run
+> `cargo dev bless`.
+
```
Comment (flip1995) [inline book/src/development/writing_tests.md]:
```suggestion
```

No longer the case.

File: book/src/development/writing_tests.md [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+Rustfix will apply the suggestions from the lint to the test file code and
+compare that to the contents of a `.fixed` file.
+
+We'll talk about suggestions more in depth in a later chapter.
+<!-- FIXME: (blyyas) Link to "Emitting lints" when that gets merged -->
+
+Use `cargo dev bless` to automatically generate the `.fixed` file after running the tests.
+
+[rustfix]: https://github.com/rust-lang/rustfix
+## Testing Manually
```
Comment (flip1995) [inline book/src/development/writing_tests.md]:
```suggestion
[rustfix]: https://github.com/rust-lang/rustfix

## Testing Manually
```

**Review by flip1995 — COMMENTED** [review]
Some more comments. I will address them right away.

You accidentally force pushed over a commit that I added to this branch. This can be avoided with force pushing with `--force-with-lease` instead of `-f`/`--force`. When using `--force-with-lease`, the force push will only go through if you fetched the latest state of the branch.

**Review by flip1995 — APPROVED** [review]

**Comment by blyxyas** [comment]
I'll merge this one first, when it gets merged #8452 will be mergeable without CI errors. Then we could fix that `FIXME` (in a later PR, currently it cannot be done as it will bug CI)

@bors r=flip1995

**Pull request merged**